File: rb-appscript/trunk/src/lib/_appscript/terminology.rb
#!/usr/local/bin/ruby
# Copyright (C) 2006 HAS. 
# Released under MIT License.

######################################################################
# TERMINOLOGY PARSER
######################################################################

module TerminologyParser
	
	require "_appscript/reservedkeywords" # names of all existing methods on ASReference::Application
	
	class BigEndianParser
		@@_nameCache = {}
		LegalFirst = 'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ_'
		LegalRest = LegalFirst + '0123456789'
		
		def initialize
			@enumerators = {}
			@properties = {}
			@commands = {}
			@_pluralClassNames = {}
			@_singularClassNames = {}
		end
		
		def _integer
			@_ptr += 2
			return @_str[@_ptr - 2, 2].unpack('S')[0]
		end
		
		def _word
			@_ptr += 4
			return @_str[@_ptr - 4, 4]
		end
		
		def _name
			count = @_str[@_ptr]
			@_ptr += 1 + count
			s = @_str[@_ptr - count, count]
			if not @@_nameCache.has_key?(s)
				legal = LegalFirst
				res = ''
				s.split(//).each do |c|
					if legal[c]
						res += c
					else
						case c
							when ' ', '-', '/'
								res += '_'
							when '&'
								res += 'and'
						else
							if res == ''
								res = '_'
							end
							res += "0x#{c.unpack('HXh')}"
						end
					end
					legal = LegalRest
				end
				if res[0, 3] == 'AS_' or ReservedKeywords.include?(res) or res[0, 1] == '_'
					res += '_'
				end
				@@_nameCache[s] = res
			end
			return @@_nameCache[s]
		end
		
		##
		
		def parseCommand
			name = _name
			@_ptr += 1 + @_str[@_ptr]
			@_ptr += @_ptr & 1
			code = _word + _word
			# skip result
			@_ptr += 4
			@_ptr += 1 + @_str[@_ptr]
			@_ptr += @_ptr & 1
			@_ptr += 2
			# skip direct
			@_ptr += 4
			@_ptr += 1 + @_str[@_ptr]
			@_ptr += @_ptr & 1
			@_ptr += 2
			#
			currentCommandArgs = []
			@commands[code] = [name, code, currentCommandArgs]
			# args
			_integer.times do
				parameterName = _name
				@_ptr += @_ptr & 1
				parameterCode = _word
				@_ptr += 4
				@_ptr += 1 + @_str[@_ptr]
				@_ptr += @_ptr & 1
				@_ptr += 2
				currentCommandArgs.push([parameterName, parameterCode])
			end
		end
		
		def parseClass
			name = _name
			@_ptr += @_ptr & 1
			code = _word
			@_ptr += 1 + @_str[@_ptr]
			@_ptr += @_ptr & 1
			isPlural = false
			_integer.times do
				propname = _name
				@_ptr += @_ptr & 1
				propcode = _word
				@_ptr += 4
				@_ptr += 1 + @_str[@_ptr]
				@_ptr += @_ptr & 1
				flags = _integer
				if propcode != 'c@#^'
					if flags & 1 == 1
						isPlural = true
					else
						@properties[propcode] = [propname, propcode]
					end
				end
			end
			_integer.times do
				@_ptr += 4
				count = _integer
				@_ptr += 4 * count
			end
			if isPlural
				@_pluralClassNames[code] = [name, code]
			else
				@_singularClassNames[code] = [name, code]
			end
		end
		
		def parseComparison
			@_ptr += 1 + @_str[@_ptr]
			@_ptr += @_ptr & 1
			@_ptr += 4
			@_ptr += 1 + @_str[@_ptr]
			@_ptr += @_ptr & 1
		end
		
		def parseEnumeration
			@_ptr += 4
			_integer.times do
				name = _name
				@_ptr += @_ptr & 1
				code = _word
				@_ptr += 1 + @_str[@_ptr]
				@_ptr += @_ptr & 1
				@enumerators[code + name] = [name, code]
			end
		end
		
		def parseSuite
			@_ptr += 1 + @_str[@_ptr]
			@_ptr += 1 + @_str[@_ptr]
			@_ptr += @_ptr & 1
			@_ptr += 4
			@_ptr += 4
			_integer.times { parseCommand }
			_integer.times { parseClass }
			_integer.times { parseComparison }
			_integer.times { parseEnumeration }
		end
		
		def parse(aetes)
			aetes.each do |aete|
				@_str = aete.data
				@_ptr = 6
				_integer.times { parseSuite }
				if not @_ptr == @_str.length
					raise RuntimeError, "aete was not fully parsed."
				end
			end
			classes = @_pluralClassNames.clone
			classes.update(@_singularClassNames)
			elements = @_singularClassNames.clone
			elements.update(@_pluralClassNames)
			return [classes, @enumerators, @properties, elements, @commands].map! { |d| d.values }
		end
	end
	
	
	class LittleEndianParser < BigEndianParser
		def _word
			return super.reverse
		end
	end
	
	
	#######
	# Public
	
	def TerminologyParser.buildTablesForAetes(aetes)
		if [1].pack('S') == "\000\001"
			return BigEndianParser.new.parse(aetes)
		else
			return LittleEndianParser.new.parse(aetes)
		end
	end

end


######################################################################
# TERMINOLOGY TABLES BUILDER
######################################################################

module Terminology

	require "aem"
	require "_appscript/defaultterminology"

	@@_terminologyCache = {}
	
	def Terminology._makeTypeTable(classes, enums, properties)
		typebycode = DefaultTerminology::TypeByCode.clone
		typebyname = DefaultTerminology::TypeByName.clone
		[[AEM::AEType, properties], [AEM::AEEnum, enums], [AEM::AEType, classes]].each do |klass, table|
			table.each do |name, code|
				if DefaultTerminology::TypeByName.has_key?(name) and \
						DefaultTerminology::TypeByName[name].code != code
					name += '_'
				end
				typebycode[code] = name.intern
				typebyname[name.intern] = klass.new(code)
			end
		end
		return [typebycode, typebyname]
	end
	
	def Terminology._makeReferenceTable(properties, elements, commands)
		referencebycode = DefaultTerminology::ReferenceByCode.clone
		referencebyname = DefaultTerminology::ReferenceByName.clone
		[[:property, properties], [:element, elements]].each do |kind, table|
			table.each do |name, code|
				referencebycode[code] = name
				referencebyname[name.intern] = [kind, code]
			end
		end
		commands.reverse.each do |name, code, args|
			if DefaultTerminology::DefaultCommands.has_key?(name) and \
					code != DefaultTerminology::DefaultCommands[name]
						name += '_'
			end
			dct = {}
			args.each { |argName, argCode| dct[argName.intern] = argCode }
			referencebyname[name.intern] = [:command, [code, dct]]
		end
		return referencebycode, referencebyname
	end
	
	#######
	# public
	
	def Terminology.defaultTables
		return _makeTypeTable([], [], []) + _makeReferenceTable([], [], [])
	end
	
	def Terminology.tablesForAetes(aetes)
		classes, enums, properties, elements, commands = TerminologyParser.buildTablesForAetes(aetes.delete_if { |aete| aete == nil or aete.type != 'aete' })
		return _makeTypeTable(classes, enums, properties) + _makeReferenceTable(properties, elements, commands)
	end
	
	##
	
	def Terminology.tablesForModule(terms)
		if terms::Version != 1.1
			raise RuntimeError, "Unsupported terminology module version: #{terms::Version} (requires version 1.1)."
		end
		return _makeTypeTable(terms::Classes, terms::Enumerators, terms::Properties) \
			+ _makeReferenceTable(terms::Properties, terms::Elements, terms::Commands)
	end
	
	def Terminology.tablesForApp(path, pid, url)
		if not @@_terminologyCache.has_key?([path, pid, url])
			begin
				if path
					app = AEM::Application.newPath(path)
				elsif pid
					app = AEM::Application.newPID(pid)
				elsif url
					app = AEM::Application.newURL(url)
				else
					app = AEM::Application.current
				end
				begin
					aetes = app.event('ascrgdte', {'----' => 0}).send(30 * 60)
				rescue AEM::CommandError => e
					if  e.number == -192 # aete resource not found
						aetes = []
					else
						raise
					end
				end
				if not aetes.is_a?(Array)
					aetes = [aetes]
				end
			rescue => err
				raise RuntimeError, "Can't get terminology for application (#{path or pid or url}): #{err}"
			end
			@@_terminologyCache[[path, pid, url]] = Terminology.tablesForAetes(aetes)
		end
		return @@_terminologyCache[[path, pid, url]]
	end
end

File: py-appscript/branches/py-appscript-0.16.x/Lib/appscript/terminology.py
#!/usr/bin/env python

"""terminology -- Obtains an application's aete resource(s) using a 'ascrgdte' event and converts them into lookup tables for use in AppData objects.

(C) 2004 HAS
"""

from aem import AEType, AEEnum, Application
from osaterminology import appscripttypedefs

from terminologyparser import buildtablesforaetes
from keywordwrapper import Keyword

__all__ = ['tablesforapp', 'tablesfordata', 'kProperty', 'kElement', 'kCommand', 'defaulttypesbycode']


######################################################################
# PUBLIC
######################################################################
# Constants

kProperty = 'p'
kElement = 'e'
kCommand = 'c'

######################################################################
# PRIVATE
######################################################################
# Cache

_terminologyCache = {} # cache parsed terminology

######################################################################
# Default terminology tables for converting between human-readable identifiers and Apple event codes; used by all apps.
# Includes default entries for Required Suite, get/set and miscellaneous other commands; application may override some or all of these definitions.

# Type tables; used to translate constants
# e.g. k.document <-> AEType('docu')
# e.g. k.ask <-> AEEnum('ask ')

_typebyname = {} # used to encode class and enumerator keywords
_typebycode = {} # used to decode class (typeType) and enumerator (typeEnum) descriptors

for _, enumerators in appscripttypedefs.enumerations:
	for name, code in enumerators:
		_typebyname[name] = AEEnum(code)
		_typebycode[code] = Keyword(name)
for defs in [
		appscripttypedefs.alltypes, 
		appscripttypedefs.commontypes, 
		appscripttypedefs.properties]:
	for name, code in defs:
		_typebyname[name] = AEType(code)
		_typebycode[code] = Keyword(name)

# Reference tables; used to translate references and commands
# e.g. app(...).documents.text <-> app.elements('docu').property('ctxt')
# e.g. app(...).quit(saving=k.ask) <-> Application(...).event('aevtquit', {'savo': AEEnum('ask ')})

_defaultcommands = {
		# 'run', 'open', 'print' and 'quit' are Required Suite commands so should always be available.
		'run': (kCommand, ('aevtoapp', {})),
		'open': (kCommand, ('aevtodoc', {})),
		'print_': (kCommand, ('aevtpdoc', {})),
		'quit': (kCommand, ('aevtquit', {'saving': 'savo'})),
		# 'reopen' and 'activate' aren't normally listed in terminology.
		'reopen': (kCommand, ('aevtrapp', {})),
		'activate': (kCommand, ('miscactv', {})),
		# 'launch' is a special case not listed in terminology. The 'real' implementation is actually in reference.Reference, as it needs to use the Process Manager to launch an application without sending it a run/open event.
		'launch': (kCommand, ('ascrnoop', {})),
		# 'get' and 'set' command often aren't listed in terminology
		'get': (kCommand, ('coregetd', {})),
		'set': (kCommand, ('coresetd', {'to': 'data'})),
		# some apps (e.g. Safari) which support GetURL events may omit it from their terminology; 'open location' is the name Standard Additions defines for this event
		'open_location': (kCommand, ('GURLGURL', {'window': 'WIND'})), 
		}

_referencebycode = { # used to decode property and element specifiers
		# some apps, e.g. Jaguar Finder, may omit 'class' property from terminology
		'ppcls': (kProperty, 'class_'),
		# some apps, e.g. iTunes, may omit 'id' property from terminology
		'pID  ': (kProperty, 'id'),
		}

_referencebyname = { # used to encode property and element specifiers and Apple events
		'class_': (kProperty, 'pcls'),
		'id': (kProperty, 'ID  '),
		}
_referencebyname.update(_defaultcommands)


######################################################################
# Translation table parsers

def _makeTypeTable(classes, enums, properties):
	# Used for constructing k.keywords
	# Each parameter is of format [[name, code], ...]
	typebycode = _typebycode.copy()
	typebyname = _typebyname.copy()
	for klass, table in [(AEType, classes), (AEEnum, enums), (AEType, properties)]: # note: packing properties as AEProp causes problems when the same name is used for both a class and a property, and the property's definition masks the class's one (e.g. Finder's 'file'); if an AEProp is passed where an AEType is expected, it can cause an error as it's not what the receiving app expects. (Whereas they may be more tolerant of an AEType being passed where an AEProp is expected.) Also, note that AppleScript always seems to pack property names as typeType, so we should be ok following its lead here.
		for name, code in table:
			# TO DO: decide where best to apply AE keyword escaping, language keyword escaping
			# TO DO: make sure same collision avoidance is done in help terminology (i.e. need to centralise all this stuff in a single osaterminology module)
			# If an application-defined name overlaps an existing type name but has a different code, append '_' to avoid collision:
			if _typebyname.has_key(name) and _typebyname[name].code != code:
				name += '_'
			typebycode[code] = Keyword(name)
			typebyname[name] = klass(code)
	return typebycode, typebyname


def _makeReferenceTable(properties, elements, commands):
	# Used for constructing references and commands
	# First two parameters are of format [[name, code], ...]
	# Last parameter is of format [name, code, direct arg type, [[arg code, arg name], ...]]
	referencebycode = _referencebycode.copy()
	referencebyname = _referencebyname.copy()
	for kind, table in [(kProperty, properties), (kElement, elements)]: # note that when an element and a property have the same name and code, we want to pack as an all-elements specifier (this is what AS does)
		for name, code in table:
			referencebycode[kind+code] = (kind, name)
			referencebyname[name] = (kind, code)
	if commands:
		for name, code, args in commands[::-1]: # if two commands have same name but different codes, only the first definition should be used (iterating over the commands list in reverse ensures this)
			# TO DO: make sure same collision avoidance is done in help terminology (i.e. need to centralise all this stuff in a single osaterminology module)
			# Avoid collisions between default commands and application-defined commands with same name but different code (e.g. 'get' and 'set' in InDesign CS2):
			if _defaultcommands.has_key(name) and code != _defaultcommands[name][1][0]:
				name += '_'
			referencebyname[name] = (kCommand, (code, dict(args)))
	return referencebycode, referencebyname


######################################################################
# PUBLIC
######################################################################

defaulttypesbycode= _typebycode # used by help system


def tablesfordata(terms):
	"""Build terminology tables from a dumped terminology module."""
	return _makeTypeTable(terms.classes, terms.enums, terms.properties) \
			+ _makeReferenceTable(terms.properties, terms.elements, terms.commands)


def tablesforapp(path=None, url=None):
	if not _terminologyCache.has_key(path or url):
		try:
			aetes = Application(path, url).event('ascrgdte', {'----':0}).send()
		except Exception, e: # (e.g.application not running)
			raise RuntimeError, "Can't get terminology for application (%s): %s" % (path or url, e)
		if not isinstance(aetes, list):
			aetes = [aetes]
		#print [aete.data for aete in aetes if aete.type == 'aete']
		classes, enums, properties, elements, commands = buildtablesforaetes([aete.data for aete in aetes if aete.type == 'aete'])
		_terminologyCache[path or url] = _makeTypeTable(classes, enums, properties) + _makeReferenceTable(properties, elements, commands)
	return _terminologyCache[path or url]


######################################################################
# TEST
######################################################################

if __name__ == '__main__':
	#for t in tablesforlocalapp('/Applications/TextEdit.app'):
	#	print t, '\n\n'
#	tablesforapp(url='eppc://mini.local/TextEdit')
#	_terminologyCache.clear()
	from time import time as t
	d=tablesforapp('/Applications/textedit.app')
	tt=t()
#	d=tablesforapp(url='eppc://mini.local/TextEdit')
	print t()-tt
	'''
	from aem import Codecs
	import InDesignCS2 as i
	c=Codecs()
	o=c.pack((i.classes, i.enums, i.properties, i.elements, i.commands))
	tt=t()
	c.unpack(o)
	d = tablesfordata(i)
	print t()-tt
	'''
	if 1:
		from pprint import pprint
		for n in d:
			pprint(n)
			print
			print